### PR TITLE
History based navigation with new right panel store

### PR DESCRIPTION
**Pull request by toger5**
Fixes: https://github.com/vector-im/element-web/issues/19686
Fixes: https://github.com/vector-im/element-web/issues/19660
Fixes: https://github.com/vector-im/element-web/issues/19634

This tires/should solve all the requirements from this post:
### The changes to watch out for in this PR:
 - complex right panel histories should be possible:
```
                                     Timeline in the right panel 
-push-a-thread-->                    Thread in the right panel
-push-a-member-profile-of-userA-->   Show member panel with userA
-call-pop/press-back-->              Thread in the right panel
-call-pop/press-back-->              Timeline in the right panel
```

### Things to test that should work:
(this pr is based on the first pr so it should include everything in this list)

-----
from the first PR:
 - [x] verification flows
 - [x] the right panel needs to be per room, changing rooms should change the right panel
 - [x] The correct member should be shown when the right panel shows a member and the room gets changed from roomA -> roomB -> roomA
 - [x] The correct Thread should be shown when the right panel shows a thread and the room gets cahnged form roomA -> roomB -> roomA
 - [x] Right panel phase should be stored on page reload (for all rooms)
 - [x] Thread right panel should be stored on page reload
 - [x] Member right panel should be stored on page reload
 - [x] Communities need to work as before (except with per room panels) (there is a back button missing in the member panel when you are viewing a community panel. Is that an issue? How long until we will drop communities. how many ppl are still using it?)

--------------
from this pr
 - [x] creating a complex right panel history in a room (A). the history should be preserved in the following scenarios
     - [x] changing rooms to room (B) and coming back to the room (A)
     - [x] restarting element by refreshing the page
     - [x] creating another complex history in a different room (C) and coming back to this room (A)
 - [x] the back button should show the correct panel when going back in the hover tooltip

[J4no wrote](https://github.com/vector-im/element-web/issues/19660#issuecomment-971574851): 
> I’m sharing here for reference the **immediate plan of action for the Threads P0 epic** that we discussed internally. We can explore more in-depth other issues and implementations outside the scope of Threads P0.
>
> - **Ideally, keep track of the status for local, room-level panels**. Local panels include the Room Info panel and the Threads panel. This way when opening a thread in a room, switching to a different room, and then going back to the first one, the thread will still be there.
> - **Only global panels, if any, can persist their contents across rooms**. We have only one global panel, the Notifications panel. That said, it’s currently barely usable as it can only show mentions in unencrypted rooms, so we could even consider not persisting it for now, and thus minimizing conflicts with any open local panel, like threads.
> - **Update the labels for the back button’s tooltip** to indicate the specific navigation destination. “Back” is ambiguous as you can’t tell in advance if it traverses navigation history (what most people will guess given the label) or the navigation hierarchy tree (what’s actually happening). Optionally, we could explore labeling the back buttons, instead of relying exclusively on tooltips.
> - **Allow for hierarchical navigation to the User Profile view from multiple contexts**: Thread context `Thread list → Thread detail → User profile`, and room info context `Room info → People list → User profile`. Otherwise, users get trapped in the user profile view when accessed in the context of a thread.

---
Here's what your changelog entry will look like:

## ✨ Features
 * History based navigation with new right panel store ([\#7398](https://github.com/matrix-org/matrix-react-sdk/pull/7398)). Fixes vector-im/element-web#19686 vector-im/element-web#19660 and vector-im/element-web#19634. Contributed by @toger5.

Preview: https://61d5c52d355fa55a73e93de2--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/structures/RoomView.tsx
Comment (gsouquet) [inline src/components/structures/RoomView.tsx:430]: How come this has been commented out? I believe this is needed when one follows a matrix.to link

Comment (toger5) [reply src/components/structures/RoomView.tsx:430]: It is commented out so that I know there is still work to do. I wanted to ask you, if the rest of the logic can also be removed here.
It is removed because (I think) its purpose was to hide/show the thread when switching rooms. Since this is now done by the store, (the store just shows the correct panel for the new room you open) it was not necessary anymore.
But thinking about it again it might also be used, to show the thread panel when using a link to an event? Than this is still needed. Can you confirm that/ explain what it does? Maybe the correct thing here would be to not remove it but change it so that it does not interfere with what the right panel store wants to do by adding another condition.

Comment (jryans) [reply src/components/structures/RoomView.tsx:430]: Yes, I believe it's needed to jump to a specific event in a thread. For example, find some event in a thread, go to message options, share, copy the message link. Now change rooms, and enter `/goto <message link>` in the composer. This should jump to the room, open the thread, and highlight the specific event.

**Review by janogarcia — APPROVED** [review]
@toger5 Thanks so much for all the work you've put into making this happen, Timo! 👏🔥

If possible, I'd recommend labelling the **Back** button tooltip dynamically in the User Info panel based on navigation history, so it reads *Back to thread* if you reached that panel from the context of a thread, or *Members* if you were viewing the Members list.

File: src/components/views/right_panel/GroupHeaderButtons.tsx
Comment (jryans) [inline src/components/views/right_panel/GroupHeaderButtons.tsx:57]: Remove...?

File: src/components/views/right_panel/RoomHeaderButtons.tsx
Comment (jryans) [inline src/components/views/right_panel/RoomHeaderButtons.tsx:161]: Remove...?

File: src/stores/right-panel/RightPanelStore.ts
Comment (jryans) [inline src/stores/right-panel/RightPanelStore.ts:175]:
```suggestion
        const history = cards.map(c => ({ phase: c.phase, state: c.state ?? {} }));
```

**Review by jryans — CHANGES_REQUESTED** [review]
Looks good overall, mainly just the deep linking to thread event to sort out, as well as any Design feedback there may be.

**Comment by janogarcia** [comment]
@toger5 🔴 **Update**: I've identified at least two blocking issues when interacting with threads.

(I'll update the description in a few minutes)

File: src/components/structures/RoomView.tsx
Comment (toger5) [reply src/components/structures/RoomView.tsx:430]: Thanks for explaingin how to force that situation (wasnt aware of `/goto`). So this code wont be commented after all. But changed.

**Comment by janogarcia** [comment]
@toger5 I've updated my previous comment on the two issues that would need to get some attention before merging.

**Review by janogarcia — CHANGES_REQUESTED** [review]
I've identified at least two blocking issues when interacting with threads.

**1. Clicking on the thread summary won't open the thread panel**

It seems you can only view a thread if you had opened previously the thread panel. You should still be able to view a thread when clicking on thread summary in the main timeline, even if the thread panel was closed before clicking on it.

**2. Clicking the back arrow on the thread detail view should take you to the All Threads view**

Just the same behavior that it's currently implemented for user profiles. For example, you can click on the main timeline on multiple avatars, one after the other, but every time the back button will be pointing to _Members_ (one level up the hierarchy three).

File: src/components/structures/RoomView.tsx
Comment (toger5) [reply src/components/structures/RoomView.tsx]: This now got changed again. and it seems everything behaves as intended now.

**Comment by toger5** [comment]
@janogarcia 
for the back button the phrasing is now as following:
 - `"Back to all threads"`
 - `"Back to chat"`
 - `"Back to info"`
 - `"Back to member list"`
 - `"Back to thread"`

**Comment by janogarcia** [comment]
@toger5 My recommendations would be:

 For the Room Information navigation path `Room info > People > User profile`: 
 
- **“Room information”** tooltip label for the back button when viewing the People list, instead of “Back to info”.
- **“Room members”** tooltip label for the back button when viewing a User Profile, instead of “Back to member list”.  

For the Thread navigation path `Thread List > Thread > User profile`:  

- **“Back to thread”** tooltip label for the back button when viewing a User Profile.  
- As for the **“Back to all threads”** label, that one shouldn’t exist in the first place. It’s a bug on the current implementation of the Thread List. I’ll let Germain know about it. Clicking anywhere on the thread tile should take you to the thread.

  Lastly, I’m not sure about  the context for the **”Back to chat”** tooltip label. Maybe it’s for the maximized widgets project Gaelle started working on?

**Comment by toger5** [comment]
Okay, thanks for the improved wording.
Yes, that is in the maximised widget context.

**Comment by janogarcia** [comment]
> Yes, that is in the maximised widget context.

@toger5 In that case I think "Back to chat" is fine.

**Review by janogarcia — CHANGES_REQUESTED** [review]
Forgot to explicitly add my review. The requested changes can be found in my previous comment: https://github.com/matrix-org/matrix-react-sdk/pull/7398#issuecomment-999679245

**Comment by toger5** [comment]
> Forgot to explicitly add my review. The requested changes can be found in my previous comment: [#7398 (comment)](https://github.com/matrix-org/matrix-react-sdk/pull/7398#issuecomment-999679245)

The wording is now as proposed.

File: src/dispatcher/dispatch-actions/threads.ts
Comment (jryans) [inline src/dispatcher/dispatch-actions/threads.ts:25]: Could we change this function to accept an object with these as properties? That way it will be clear what each value is at the call site, and unused optional values can just be omitted.

**Review by jryans — CHANGES_REQUESTED** [review]
Overall it's coming together, just one tweak that should make it more readable going forward. 😄

File: src/dispatcher/dispatch-actions/threads.ts
Comment (toger5) [reply src/dispatcher/dispatch-actions/threads.ts:25]: I was going back and forth with this. Its good to have a second opinion. When using it more I also had the feeling that having named params would make everything clearer.

File: src/components/structures/RoomView.tsx
Comment (jryans) [inline src/components/structures/RoomView.tsx:428]: When the key and value are the same name, you don't need to write both:

```suggestion
                    initialEvent,
```

File: src/components/structures/RoomView.tsx
Comment (jryans) [inline src/components/structures/RoomView.tsx:438]:
```suggestion
                        initialEvent,
```

File: src/components/views/right_panel/EncryptionPanel.tsx
Comment (jryans) [inline src/components/views/right_panel/EncryptionPanel.tsx:125]: Either no braces or move to the next line and indent.

```suggestion
        if (!RightPanelStore.instance.isOpenForRoom) RightPanelStore.instance.togglePanel();
```

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (jryans) [inline src/components/views/toasts/VerificationRequestToast.tsx:125]:
```suggestion
                RightPanelStore.instance.setCards(
                    [
                        { phase: RightPanelPhases.RoomSummary },
                        { phase: RightPanelPhases.RoomMemberInfo, state: { member } },
                        { phase: RightPanelPhases.EncryptionPanel, state: { verificationRequest: request, member } },
                    ],
                    undefined,
                    request.channel.roomId,
                );
```

File: src/verification.ts
Comment (jryans) [inline src/verification.ts:114]:
```suggestion
    if (RightPanelStore.instance.roomPhaseHistory.some((card) => (card.phase == RightPanelPhases.RoomSummary))) {
```

**Review by jryans — CHANGES_REQUESTED** [review]
Looks good overall, just a few style cleanups to apply. 😄

File: src/stores/right-panel/RightPanelStoreIPanelState.ts
Comment (jryans) [inline src/stores/right-panel/RightPanelStoreIPanelState.ts:116]: The `as any` at the top of both of these conversion functions also seems surprising, can we remove those as well?

**Review by jryans — CHANGES_REQUESTED** [review]

File: src/stores/right-panel/RightPanelStoreIPanelState.ts
Comment (toger5) [reply src/stores/right-panel/RightPanelStoreIPanelState.ts:116]: I think we cant:
```
const panelStateThisRoom = { ...panelStateStore?.state } as any;
    if (!!panelStateThisRoom.threadHeadEventId) {
        panelStateThisRoom.threadHeadEvent = room.findEventById(panelStateThisRoom.threadHeadEventId);
    }
```
The panelStateStore.state does not have the property `threadHeadEvent` it only has `threadHeadEventId`. So the linter would be mad.

Comment (toger5) [reply src/stores/right-panel/RightPanelStoreIPanelState.ts:116]: I just tried without `as any` and with `as IRightPanelCardState` both cause the linter to be unhappy. Any other way to make it less surprising?

**Review by janogarcia — CHANGES_REQUESTED** [review]
@toger5 Thanks for implementing all the suggested changes, Timo! Just one minor tweak, please change the copy for `Back to all threads` to just `Threads`.

**Review by janogarcia — APPROVED** [review]
👏

File: src/stores/right-panel/RightPanelStoreIPanelState.ts
Comment (jryans) [reply src/stores/right-panel/RightPanelStoreIPanelState.ts:116]: Okay, I'll check this out locally and see if there's a better way.

Comment (jryans) [reply src/stores/right-panel/RightPanelStoreIPanelState.ts:116]: Perhaps it's better to explicitly copy the safe keys at first, and then you don't need the `delete` lines either...?

```
export function convertCardToStore(panelState: IRightPanelCard): IRightPanelCardStored {
    const panelStateThisRoomStored: IRightPanelCardStateStored = {
        groupId: panelState?.state?.groupId,
        groupRoomId: panelState?.state?.groupRoomId,
        widgetId: panelState?.state?.widgetId,
        spaceId: panelState?.state?.spaceId,
        isInitialEventHighlighted: panelState?.state?.isInitialEventHighlighted,
    };
    if (!!panelState?.state?.threadHeadEvent?.getId()) {
        panelStateThisRoomStored.threadHeadEventId = panelState.state.threadHeadEvent.getId();
    }
    if (!!panelState?.state?.memberInfoEvent?.getId()) {
        panelStateThisRoomStored.memberInfoEventId = panelState.state.memberInfoEvent.getId();
    }
    if (!!panelState?.state?.initialEvent?.getId()) {
        panelStateThisRoomStored.initialEventId = panelState.state.initialEvent.getId();
    }
    if (!!panelState?.state?.member?.userId) {
        panelStateThisRoomStored.memberId = panelState.state.member.userId;
    }

    const storedCard = { state: panelStateThisRoomStored, phase: panelState.phase };
    return storedCard;
}
```

Also at the moment, this guards against every possible level being undefined... but don't we know that at least `panelState` and `panelState.state` should exist, and those can use regular `.` instead of `?.`?

**Review by jryans — APPROVED** [review]
Assuming the CI errors are resolved, this looks good to me! 😄

**Review by jryans — APPROVED** [review]

**Pull request merged**
